File: 2023/lang-rust/src/day02.rs
use regex::Regex;
use std::fs;

fn part_1(input: &str) -> u32 {
    let regex_game = Regex::new(r"Game (\d+): (.*)").unwrap();

    input
        .lines()
        .map(|line: &str| {
            let re_match = regex_game.captures(line).unwrap();
            let game = re_match.get(1).unwrap().as_str().parse::<u32>().unwrap();
            let rounds: Vec<&str> = re_match.get(2).unwrap().as_str().split(";").collect();
            let mut rounds_iter = rounds.into_iter();
            let mut is_invalid: bool = false;

            while let Some(round) = rounds_iter.next() {
                let items: Vec<&str> = round.trim().split_whitespace().collect();
                let mut items_iter = items.into_iter();

                while let Some(count) = items_iter.next() {
                    let total = count.parse::<u32>().unwrap();
                    let color = items_iter.next().unwrap().trim_end_matches(",");

                    if color == "red" && total > 12 {
                        is_invalid = true;
                    }
                    if color == "green" && total > 13 {
                        is_invalid = true;
                    }
                    if color == "blue" && total > 14 {
                        is_invalid = true;
                    }
                }
            }

            if is_invalid {
                0
            } else {
                game
            }
        })
        .sum()
}

fn part_2(input: &str) -> u32 {
    let regex_game = Regex::new(r"Game (\d+): (.*)").unwrap();

    input
        .lines()
        .map(|line: &str| {
            let mut color_r: u32 = 0;
            let mut color_g: u32 = 0;
            let mut color_b: u32 = 0;

            let re_match = regex_game.captures(line).unwrap();
            let game = re_match.get(1).unwrap().as_str().parse::<u32>().unwrap();
            let rounds: Vec<&str> = re_match.get(2).unwrap().as_str().split(";").collect();
            let mut rounds_iter = rounds.into_iter();

            while let Some(round) = rounds_iter.next() {
                let items: Vec<&str> = round.trim().split_whitespace().collect();
                let mut items_iter = items.into_iter();

                while let Some(count) = items_iter.next() {
                    let total = count.parse::<u32>().unwrap();
                    let color = items_iter.next().unwrap().trim_end_matches(",");

                    if color == "red" {
                        color_r = color_r.max(total);
                    }
                    if color == "green" {
                        color_g = color_g.max(total);
                    }
                    if color == "blue" {
                        color_b = color_b.max(total);
                    }
                }
            }

            color_r * color_g * color_b
        })
        .sum()
}

pub fn solve() {
    let input = fs::read_to_string("../input/day02.txt").expect("Missing");

    println!("Day 02, Part 1 = {}", part_1(&input));
    println!("Day 02, Part 2 = {}", part_2(&input));
}

File: 2023/lang-php/src/Day02.php
<?php

class Day02
{
    public static function part1(string $input): int
    {
        $lines = explode(string: trim($input), separator: \PHP_EOL);
        $output = 0;

        foreach ($lines as $line) {
            $line_split = explode(string: $line, separator: ':');
            $game_id = intval(substr(string: $line_split[0] ?? '', offset: 5));
            $invalid = false;

            $rounds = trim(string: $line_split[1] ?? '');
            $rounds = explode(string: $rounds, separator: ';');
            $rounds = array_map(array: $rounds, callback: fn ($round) => explode(string: trim($round), separator: ','));

            foreach ($rounds as $round) {
                $options = array_map(array: $round, callback: fn ($option) => explode(string: trim($option), separator: ' '));

                foreach ($options as $option) {
                    $count = intval($option[0]);
                    $color = $option[1];

                    $invalid = match (true) {
                        $color === 'red' && $count > 12 => true,
                        $color === 'green' && $count > 13 => true,
                        $color === 'blue' && $count > 14 => true,
                        default => $invalid,
                    };
                }
            }

            if ($invalid === false) {
                $output += $game_id;
            }
        }

        return $output;
    }

    public static function part2(string $input): int
    {
        $lines = explode(string: trim($input), separator: \PHP_EOL);
        $output = 0;

        foreach ($lines as $line) {
            $line_split = explode(string: $line, separator: ':');
            $min_values = ['red' => 0, 'green' => 0, 'blue' => 0];

            $rounds = trim(string: $line_split[1] ?? '');
            $rounds = explode(string: $rounds, separator: ';');
            $rounds = array_map(array: $rounds, callback: fn ($round) => explode(string: trim($round), separator: ','));

            foreach ($rounds as $round) {
                $options = array_map(array: $round, callback: fn ($option) => explode(string: trim($option), separator: ' '));

                foreach ($options as $option) {
                    $count = intval($option[0]);
                    $color = $option[1];

                    $min_values[ $color ] = max($min_values[ $color ], intval($count));
                }
            }

            $output += ($min_values['red'] * $min_values['green'] * $min_values['blue']);
        }

        return $output;
    }
}

$input = file_get_contents(filename: '../input/day02.txt');
echo sprintf("Day 02, Part 01 = %d", Day02::part1($input)) . \PHP_EOL;
echo sprintf("Day 02, Part 02 = %d", Day02::part2($input)) . \PHP_EOL;
